src/vm/inlinetracking.cpp

```diff
@@ -418,12 +424,181 @@ void PersistentInlineTrackingMap::Save(DataImage *image, InlineTrackingMap* runt
         m_inlineeIndexSize * sizeof(m_inlineeIndex[0]), m_inlinersBufferSize));
 }
 
-void PersistentInlineTrackingMap::Fixup(DataImage *image)
+void PersistentInlineTrackingMapNGen::Fixup(DataImage *image)
+{
+    STANDARD_VM_CONTRACT;
+    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMapNGen, m_module));
+    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMapNGen, m_inlineeIndex));
+    image->FixupPointerField(this, offsetof(PersistentInlineTrackingMapNGen, m_inlinersBuffer));
+}
+
+#endif //FEATURE_NATIVE_IMAGE_GENERATION
+
+struct InliningHeader
+{
+    short MajorVersion;
+    short MinorVersion;
+    int SizeOfInlineeIndex;
+};
+
+#ifdef FEATURE_NATIVE_IMAGE_GENERATION
+
+void PersistentInlineTrackingMapR2R::ProcessInlineTrackingEntry(SBuffer *inlinersBuffer, SArray<InlineeRecordR2R> *inlineeIndex, InlineTrackingEntry *entry)
+{
+    STANDARD_VM_CONTRACT;
+    // This call removes duplicates from inliners and makes sure they are sorted
+    entry->SortAndDeduplicate();
+    MethodInModule inlinee = entry->m_inlinee;
+    InlineSArray<MethodInModule, 3> &inliners = entry->m_inliners;
+    COUNT_T tatalInlinersCount = inliners.GetCount();
+    _ASSERTE(tatalInlinersCount > 0);
+
+    // Saving number of inliners
+    NibbleWriter inlinersStream;
+    inlinersStream.WriteEncodedU32(tatalInlinersCount);
+
+    // Saving inliners RIDs, each new RID is represented as an adjustment (diff) to the previous one
+    RID prevMethodRid = 0;
+    for (COUNT_T i = 0; i < tatalInlinersCount; i++)
+    {
+        RID methodRid = RidFromToken(inliners[i].m_methodDef);
+        _ASSERTE(methodRid >= prevMethodRid);
+        inlinersStream.WriteEncodedU32(methodRid - prevMethodRid);
+        prevMethodRid = methodRid;
+    }
+    inlinersStream.Flush();
+
+    // Copy output of NibbleWriter into a big buffer (inlinersBuffer) for inliners
+    // and create an InlineeRecord with correct offset
+    InlineeRecordR2R record(RidFromToken(inlinee.m_methodDef));
+    DWORD inlinersStreamSize;
+    const BYTE *inlinersStreamPtr = (const BYTE *)inlinersStream.GetBlob(&inlinersStreamSize);
+    record.m_offset = inlinersBuffer->GetSize();
+    inlinersBuffer->Insert(inlinersBuffer->End(), SBuffer(SBuffer::Immutable, inlinersStreamPtr, inlinersStreamSize));
+
+    inlineeIndex->Append(record);
+}
+
+void PersistentInlineTrackingMapR2R::Save(SBuffer* pSaveTarget, InlineTrackingMap* runtimeMap)
```

Comment:
>Save [](start = 37, length = 4)

Nit: Ideally, we should share the pre-saving logic between this R2R and the NGEN persistent inline tracking map, and just have the Save method do the acutal writing (the pre-saving logic is the same in both case)